### ISSUE TITLE
Missing support for set-src/set-src-port for `tcp-request content`

### ISSUE THREAD

**Issue by anigeo** [haproxy/haproxy#1303]
## Detailed description of the problem

I'm trying to have a lua script to generate source ip with using `core.register_fetches` feature, while this works on HTTP mode with `http-request set-src`, but it doesn't work with `tcp-request connection set-src` in both TCP or HTTP mode

## Expected behavior

The registered fetches should be triggered by `tcp-request connection set-src` as well

## Steps to reproduce the behavior

Use the config below and try to access, while without calling lua it is working fine with `curl --interface 2001:db8::4 localhost` :
```
Jun 23 09:16:05 redacted haproxy[17645]: 00000000:test.accept(0007)=001f from [2001:db8::beef:58085] ALPN=<none>
Jun 23 09:16:05 redacted haproxy[17645]: 00000000:test.srvcls[001f:adfd]
Jun 23 09:16:05 redacted haproxy[17645]: 00000000:test.clicls[001f:adfd]
Jun 23 09:16:05 redacted haproxy[17645]: 00000000:test.closed[001f:adfd]
Jun 23 09:16:05 redacted haproxy[17645]: 2001:db8::beef:58085 [23/Jun/2021:09:16:05.109] test test/upstream 1/0/0 195 -- 1/1/0/0/0 0/0
```

But after swapping `ipv6()` with `lua.set_source_ip`, it doesn't work anymore, and it doesn't throw any error

```
Jun 23 09:16:57 redacted haproxy[17661]: 00000000:test.accept(0007)=001f from [2001:db8::4:59825] ALPN=<none>
Jun 23 09:16:57 redacted haproxy[17661]: 00000000:test.srvcls[001f:adfd]
Jun 23 09:16:57 redacted haproxy[17661]: 00000000:test.clicls[001f:adfd]
Jun 23 09:16:57 redacted haproxy[17661]: 00000000:test.closed[001f:adfd]
Jun 23 09:16:57 redacted haproxy[17661]: 2001:db8::4:59825 [23/Jun/2021:09:16:57.853] test test/upstream 1/0/0 192 -- 1/1/0/0/0 0/0
```

## Do you have any idea what may have caused this?

FYI, I've tried to use `converter` or `action` with `set-var` to archive similar goal, but result still the same, it looks to me lua has been completely ignored by `set-src` under tcp mode

## Do you have an idea how to solve the issue?

N/A

## What is your configuration?

Simplified lua script:

```lua
local function set_source_ip(txn)
        core.Alert('fetch ' .. txn.f:src())
        return '2001:db8:1234::beef'
end

core.register_fetches("set_source_ip", set_source_ip)
```

Sample haproxy.cfg:

```
global
        lua-load /etc/haproxy/test.lua

listen test
        mode tcp
        bind ipv6@:80 v6only
        source 0.0.0.0 usesrc clientip
        # tcp-request connection set-src ipv6(2001:db8::beef)  # This work
        tcp-request connection set-src lua.set_source_ip    # This doesn't
        server upstream ::1:8080
```

## Output of `haproxy -vv` and `uname -a`

```
Linux redacted 5.10.38-1-lts #1 SMP Wed, 19 May 2021 08:46:06 +0000 x86_64 GNU/Linux
```

```
HA-Proxy version 2.2.9-1 2021/02/06 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2025.
Known bugs: http://www.haproxy.org/bugs/bugs-2.2.9.html
Running on: Linux 5.10.38-1-lts #1 SMP Wed, 19 May 2021 08:46:06 +0000 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = gcc
  CFLAGS  = -O2 -g -O2 -ffile-prefix-map=/build/haproxy-KCirew/haproxy-2.2.9=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -Wall -Wextra -Wdeclaration-after-statement -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-stringop-overflow -Wno-cast-function-type -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_OPENSSL=1 USE_LUA=1 USE_ZLIB=1 USE_SYSTEMD=1
  DEBUG   =

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 -CLOSEFROM +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=8).
Built with OpenSSL version : OpenSSL 1.1.1i  8 Dec 2020
Running on OpenSSL version : OpenSSL 1.1.1k  25 Mar 2021
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.3
Built with network namespace support.
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.36 2020-12-04
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 10.2.1 20210110
Built with the Prometheus exporter as a service

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
            fcgi : mode=HTTP       side=BE        mux=FCGI
       <default> : mode=HTTP       side=FE|BE     mux=H1
              h2 : mode=HTTP       side=FE|BE     mux=H2
       <default> : mode=TCP        side=FE|BE     mux=PASS

Available services : prometheus-exporter
Available filters :
	[SPOE] spoe
	[COMP] compression
	[TRACE] trace
	[CACHE] cache
	[FCGI] fcgi-app
```

## If HAProxy crashed: Last outputs and backtraces

N/A

**Comment by capflam** [comment]
It is expected. Internally, Lua fetches/converters/actions/services depend on streams. It is a technical limitation. The lua thread is created when the stream is created. Thus, it is not possible to call a sample fetch or a converter from `tcp-request connection` and `tcp-request session` rules because the stream is created after the evaluation of these rulesets. However, from a `tcp-request content` rule, it should work.

**Comment by xv-georgef** [comment]
Thanks @capflam, meanwhile `set-src` action is not available with `tcp-request content`, does that mean there's no way to have a dynamically assigned source address?

Ref:
https://github.com/haproxy/haproxy/blob/b28430591d18f7fda5bac2e0ea590b3a34f04601/src/tcp_act.c#L320-L325

**Comment by anigeo** [comment]
Thanks @capflam, meanwhile `set-src` action is not available with `tcp-request content`, does that mean there's no way to have a dynamically assigned source address?

Ref:
https://github.com/haproxy/haproxy/blob/b28430591d18f7fda5bac2e0ea590b3a34f04601/src/tcp_act.c#L320-L325

**Comment by capflam** [comment]
Hum indeed. And it is a bit strange. Because this action is available for `http-request` rules.

**Comment by anigeo** [comment]
So I just did a quick and dirty test, apparently this can be resolve by adding `set-src` action into `static struct action_kw_list tcp_req_cont_actions` (testing with 2.2.9 debian's source)

```diff
--- haproxy-2.2.9.orig/src/proto_tcp.c
+++ haproxy-2.2.9/src/proto_tcp.c
@@ -2001,6 +2001,8 @@ static struct action_kw_list tcp_req_ses
 INITCALL1(STG_REGISTER, tcp_req_sess_keywords_register, &tcp_req_sess_actions);

 static struct action_kw_list tcp_req_cont_actions = {ILH, {
+	{ "set-src",      tcp_parse_set_src_dst },
+	{ "set-src-port", tcp_parse_set_src_dst },
 	{ "set-dst"     , tcp_parse_set_src_dst },
 	{ "set-dst-port", tcp_parse_set_src_dst },
 	{ "silent-drop",  tcp_parse_silent_drop },
 ```
 
I wonder if this will cause any unexpected behaviour.

**Comment by capflam** [comment]
Yes, I'll add the support for these actions for `tcp-request content` rules.

**Comment by capflam** [comment]
I updated the issue title because for me it is the true issue here. I hope it's ok for you. The fix was merged in upstream and will be backported soon.

**Comment by anigeo** [comment]
Much appreciated for the quick fix, and the title update.

**Comment by capflam** [comment]
Unfortunately, I reverted the fix because there is a deeper problem by setting source address at the stream level. See #90 for details, especially last [Willy's comment](https://github.com/haproxy/haproxy/issues/90#issuecomment-872298096). The feature is still mandatory and we will work on it. But we must be careful the situation does not get worse.

**Comment by kornaz** [comment]
@anigeo I needed this to work so I came up with a workaround. Add another listener and in the first listener instead of "connection set-src", do "content set-dst" (change the destination address, not the source; also, Lua works in "content" rules). Then forward everything to the second listener and do "connection set-src" with a "dst" fetch. Basically this listener is needed to copy the dst address to src and proceed further. I've modified your example.
```
listen test
        mode tcp
        bind ipv6@:80 v6only
        # tcp-request connection set-src ipv6(2001:db8::beef)  # This work
        tcp-request content set-dst lua.set_source_ip
        tcp-request content set-dst-port int(10080)
        server self :::0

listen dst-to-src
        bind ipv6@:10080
        mode tcp
        tcp-request connection set-src dst
        server upstream ::1:8080 source 0.0.0.0 usesrc clientip
```

**Comment by wtarreau** [comment]
Thanks for sharing, @kornaz. For now we still have to deal with a large number of issues that keeps everyone busy (especially during the vacation period) but I absolutely want to have this solved in 2.5 and if possible backported.

**Comment by capflam** [comment]
FYI, the patch was merged again to the 2.5 thanks to recent changes.

**Comment by capflam** [comment]
I'm closing the issue. Note that the patch cannot be backported because it relies on features only available in the 2.5.
